### PR TITLE
System Stat plugin

### PR DESCRIPTION
**Pull request by Hzj-jie**

### CODE REVIEW

**Comment by Hzj-jie** [comment]
We can merge memory usage and storage statistics into this plugin.

File: plugins/systemstat.koplugin/main.lua [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+function SystemStat:onNotCharging()
+    self.decharging_count = self.decharging_count + 1
+end
+
+function SystemStat:onCallback()
+    local kv_pairs = {
+        {_("Startup at"), os.date("%c", self.start_sec)},
+        {_("Running hours"),
+         string.format("%.2f", os.difftime(os.time(), self.start_sec) / 60 / 60)},
+        {_("Wakeup times"), self.wakeup_count},
```
Comment (Frenzie) [inline plugins/systemstat.koplugin/main.lua]: "Number of wake-ups" or "Wake-up count" (just like the variable).

File: plugins/systemstat.koplugin/main.lua [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+local UIManager = require("ui/uimanager")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local _ = require("gettext")
+
+local SystemStat = {
+    start_sec = os.time(),
+    wakeup_count = 0,
+    sleep_count = 0,
+    charging_count = 0,
+    decharging_count = 0,
```
Comment (Frenzie) [inline plugins/systemstat.koplugin/main.lua]: discharge, not decharge

File: plugins/systemstat.koplugin/main.lua [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+end
+
+function SystemStat:onNotCharging()
+    self.decharging_count = self.decharging_count + 1
+end
+
+function SystemStat:onCallback()
+    local kv_pairs = {
+        {_("Startup at"), os.date("%c", self.start_sec)},
+        {_("Running hours"),
```
Comment (Frenzie) [inline plugins/systemstat.koplugin/main.lua]: How geeky is uptime? How about "Operating time"? I'm just not sure how clear "running hours" is. What does Win/Mac say?

File: plugins/systemstat.koplugin/main.lua [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    self.charging_count = self.charging_count + 1
+end
+
+function SystemStat:onNotCharging()
+    self.decharging_count = self.decharging_count + 1
+end
+
+function SystemStat:onCallback()
+    local kv_pairs = {
+        {_("Startup at"), os.date("%c", self.start_sec)},
```
Comment (Frenzie) [inline plugins/systemstat.koplugin/main.lua]: Maybe "KOReader started at"? As phrased it makes me thing of when the system started.

**Comment by AlanSP1** [comment]
Just to mention, this looks great and have great promises. :+1:

File: plugins/systemstat.koplugin/main.lua (hunk 2 above)
Comment (Hzj-jie) [inline plugins/systemstat.koplugin/main.lua]: Done ...

File: plugins/systemstat.koplugin/main.lua (hunk 4 above)
Comment (Hzj-jie) [inline plugins/systemstat.koplugin/main.lua]: Done.

File: plugins/systemstat.koplugin/main.lua (hunk 3 above)
Comment (Hzj-jie) [inline plugins/systemstat.koplugin/main.lua]: Android uses "Up time". We may use "Up hours".

File: plugins/systemstat.koplugin/main.lua (hunk 1 above)
Comment (Hzj-jie) [inline plugins/systemstat.koplugin/main.lua]: Done.

**Comment by houqp** [comment]
Yeah, merging those two plugins into this one is a good idea :+1:

File: plugins/systemstat.koplugin/main.lua [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+function SystemStat:onCharging()
+    self.charging_count = self.charging_count + 1
+end
+
+function SystemStat:onNotCharging()
+    self.discharging_count = self.discharging_count + 1
+end
+
+function SystemStat:onCallback()
```
Comment (houqp) [inline plugins/systemstat.koplugin/main.lua]: nitpick, i suggest renaming `onCallback` to something like `showStatistics` for two reasons:

1. The coding convention is to use `onFoo` to define a handler for event `Foo`. In this case, there is no `Callback` event.
2. `onCallback` is such a generic name that it's hard to tell what this function actually does without looking at the implementation.

File: plugins/systemstat.koplugin/main.lua [hunk 6]
```diff
@@ -25,18 +25,17 @@ function SystemStat:onCharging()
...
-         string.format("%.2f", os.difftime(os.time(), self.start_sec) / 60 / 60)},
-        {_("Wakeup times"), self.wakeup_count},
-        {_("Sleep times"), self.sleep_count},
-        {_("Charging times"), self.charging_count},
-        {_("Decharging times"), self.decharging_count},
+        {_("KOReader Started at"), os.date("%c", self.start_sec)},
+        {_("Up hours"), string.format("%.2f", os.difftime(os.time(), self.start_sec) / 60 / 60)},
+        {_("Number of wake-ups"), self.wakeup_count},
+        {_("Number of sleeps"), self.sleep_count},
+        {_("Number of chargings"), self.charging_count},
```
Comment (Frenzie) [inline plugins/systemstat.koplugin/main.lua]: `Number of charges` (and discharges) would be better than this gerund. Alternatively `Number of charging cycles` would be acceptable, but even then I'd say `Number of charge cycles` is preferable.[1]

Also, I'd personally call the variables `charge_count` and `discharge_count`, but that's a total nitpick.

[1] Apple uses that, for example.
>All rechargeable batteries have a limited number of charge cycles and may eventually need to be serviced. 
Source: http://www.apple.com/batteries/

File: frontend/ui/elements/filemanager_menu_order.lua [hunk 7]
```diff
@@ -56,6 +56,7 @@ local order = {
         "version",
         "help",
         "----------------------------",
+        "system_statistics",
```
Comment (Frenzie) [inline frontend/ui/elements/filemanager_menu_order.lua]: Incidentally, does system_statistics go with exit? I'm inclined to think it's either its own thing or part of version and help. Then again, some programs (Opera/Blink) group help and about with exit.

Comment (houqp) [inline frontend/ui/elements/filemanager_menu_order.lua]: agree, part of version section makes more sense.

**Comment by houqp** [comment]
this should close #2683

File: frontend/ui/elements/filemanager_menu_order.lua (hunk 7 above)
Comment (Hzj-jie) [inline frontend/ui/elements/filemanager_menu_order.lua]: It looks like I have missed this comment.

File: plugins/systemstat.koplugin/main.lua (hunk 5 above)
Comment (Hzj-jie) [inline plugins/systemstat.koplugin/main.lua]: Updated.

File: plugins/systemstat.koplugin/main.lua (hunk 6 above)
Comment (Hzj-jie) [inline plugins/systemstat.koplugin/main.lua]: Updated.

**Pull request merged**